src/rust/engine/process_execution/src/lib.rs

```diff
@@ -410,17 +410,29 @@ pub struct ProcessResultMetadata {
   pub total_elapsed: Option<Duration>,
   /// The source of the result.
   pub source: ProcessResultSource,
+  /// The RunId of the Session in which the `ProcessResultSource` was accurate. In further runs
+  /// within the same process, the source of the process implicitly becomes memoization.
+  pub source_run_id: u32,
 }
 
 impl ProcessResultMetadata {
-  pub fn new(total_elapsed: Option<Duration>, source: ProcessResultSource) -> Self {
-    ProcessResultMetadata {
+  pub fn new(
+    total_elapsed: Option<Duration>,
+    source: ProcessResultSource,
+    context: &Context,
```

Comment:
A Context is pretty fat. I'm guessing you considered a newtype for run_id though and rejected that?

Reply:
Yea, but I dismissed it initially because it needed to be an `AtomicU32` in a few contexts... probably still worth it though. Will give it a swing.